Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_ARRAY_SORTED_MAP_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_ARRAY_SORTED_MAP_H_
+
+#include <assert.h>
+
+#include <algorithm>
+#include <array>
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/immutable/map_entry.h"
+
+namespace firebase {
+namespace firestore {
+namespace immutable {
+
+/**
+ * A base class for implementing ArraySortedMap, which contains constants and
+ * types that don't depend upon the type of ArraySortedMap's type parameters.
+ */
+class ArraySortedMapBase {
```

Comment:
Ah yes. While rebasing my next changes on top of this feedback I now realize why the base class is preferable.

In the next changes that implement `find_index`, they return an `npos` constant in the case where the key is not found. Unlike `kFixedSize` which only is used in implementation, `npos` must be a part of the interface.

Using aliases work for types but not for constants. So that technique doesn't help here.

Ultimately the problem to solve is that if we do nothing, this declaration:
```c++
template <typename K, typename V, typename C = std::less<K>>
class ArraySortedMap {
 public:
  static constexpr size_type npos = impl::npos;
};
```
Generates the following linker error:
```
Undefined symbols for architecture x86_64:
  "firebase::firestore::immutable::ArraySortedMap<int, int, std::__1::less<int> >::npos", referenced from:
      firebase::firestore::immutable::ArraySortedMap_FindIndex_Test::TestBody() in array_sorted_map_test.cc.o
```
Declaring `npos` in a public base class allowed it to be defined once rather than once per instantiation of the template.

Unless you can think of something better than the public base class I'll change this back.

Reply:
Ah, I see, thanks for explaining. User code could access npos from a namespace, but having it the class is certainly more straightforward, and also consistent with what std::string does.